### ISSUE TITLE
All `once` handlers are being flagged as used as soon as a request is intercepted, no matter the URL

### ISSUE THREAD

**Issue by acidoxee** [kettanaito/msw#1782]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Node.js version

v18.18.0

### Reproduction repository

https://stackblitz.com/edit/vitest-dev-vitest-xa3evs?file=test%2Fbasic.test.ts

### Reproduction steps

Just let the tests run (which should happen as soon as you land on the page), the green test highlights that it shouldn't behave that way.

### Current behavior

I've got a [detailed answer here already](https://github.com/mswjs/msw/pull/1436#issuecomment-1773370548), but for good measure:

- register a `once` handler, for instance for `https://test.local/foo/trpc/getById`
- perform a request for another URL, for instance for `https://test.local/bar/trpc/getByName`
- perform a request for the supposedly mocked URL (`https://test.local/foo/trpc/getById`), see that it's not properly handled and that the `once` handler had already been flagged as used

I've seen this behavior in `0.0.0-fetch.rc-23` and `0.0.0-fetch.rc-24`.

### Expected behavior

The `once` handler registered for `https://test.local/foo/trpc/getById` should not have been flagged as used after the first request, because it didn't match. In the reproduction repo, there should even still be one `once` unused handler in the pool, since only one request to `https://test.local/foo/trpc/getById` has been triggered, whereas there were two `once` handlers registered for that URL.

**Comment by kettanaito** [comment]
Thanks for reporting this, @acidoxee! I'm trying to reproduce this behavior in #1781, so far without luck. The tests are passing, and non-matching handlers are not marked as used.

**Comment by acidoxee** [comment]
It seems you're using lower-level primitives to test the behavior, whereas in my reproduction I'm using plain `fetch` requests. Maybe there's something wrong in-between? Sorry if that doesn't make sense, I'm not familiar with MSW's internals at all 🙈

Or maybe you're using a different Node version? I'm having the problem on `18.17.1`, `18.18.0` and `18.18.2`, I haven't tried with other versions.

I'm also using Vitest, not Jest, both in my own repo and in the reproduction one. I don't have any idea whether that could influence the results.

**Comment by joshkel** [comment]
I ran into this today.

It seems that the problem is here:

https://github.com/mswjs/msw/blob/88c45ea5295fa7a47a943311dff65c99ebb3111e/src/core/handlers/RequestHandler.ts#L193-L210

It immediately marks the handler as used, to prevent two back-to-back requests from both going through. However, parsing the request, to determine if the handler should be invoked, is asynchronous (as needed by the GraphQL handler). There appears to be no way to have both "immediately enforce 'once'" and "asynchronously determine whether it's used", unless you do something such as changing `isUsed` to a promise or serializing access to it via a mutex?

Here's a test in `handleRequest.test.ts` that should suffice to reproduce the issue.

<details>
<summary>(collapsed for a shorter issue description)</summary>

```ts
it('does not mark non-matching one-time handlers as used', async () => {
  const { emitter } = setup()

  const oneTimeHandler = http.get(
    '/resource',
    () => {
      return HttpResponse.text('One-time')
    },
    { once: true },
  )
  const anotherHandler = http.get(
    '/another',
    () => {
      return HttpResponse.text('Another')
    },
    { once: true },
  )
  const handlers: Array<RequestHandler> = [oneTimeHandler, anotherHandler]

  const requestId = uuidv4()
  const firstResult = await handleRequest(
    new Request('http://localhost/another'),
    requestId,
    handlers,
    options,
    emitter,
    callbacks,
  )

  expect(await firstResult?.text()).toBe('Another')
  expect(oneTimeHandler.isUsed).toBe(false)
  expect(anotherHandler.isUsed).toBe(true)

  const secondResult = await handleRequest(
    new Request('http://localhost/resource'),
    requestId,
    handlers,
    options,
    emitter,
    callbacks,
  )

  expect(await secondResult?.text()).toBe('One-time')
  expect(anotherHandler.isUsed).toBe(true)
  expect(oneTimeHandler.isUsed).toBe(true)
})
```

**Comment by kettanaito** [comment]
## Released: v2.0.2 🎉

This has been released in v2.0.2!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.0.2)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.0.2)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
